server/app/services/geo/esri/RealEsriClient.java

```diff
@@ -99,23 +112,117 @@ private CompletionStage<WSResponse> tryRequest(WSRequest request, int retries) {
   @Override
   @VisibleForTesting
   CompletionStage<Optional<JsonNode>> fetchAddressSuggestions(ObjectNode addressJson) {
-    if (this.ESRI_FIND_ADDRESS_CANDIDATES_URL.isEmpty()) {
+    // Check if the URL list is empty
+    if (ESRI_FIND_ADDRESS_CANDIDATES_URLS.isEmpty()) {
       return CompletableFuture.completedFuture(Optional.empty());
     }
-    WSRequest request = ws.url(this.ESRI_FIND_ADDRESS_CANDIDATES_URL.get());
+
+    return processAddressSuggestionUrlsSequentially(
+        addressJson, ESRI_FIND_ADDRESS_CANDIDATES_URLS, Optional.empty());
+  }
+
+  /**
+   * Recursive method to get address correction data from one or more findAddressCandidates
+   * endpoints. This will stop when there are no urls in the list to process or we end early because
+   * we have results that are greater than the scoring threashold.
+   *
+   * @param addressJson Uncorrected address
+   * @param urls A list of urls that will be processed in order; may stop early if surpassing score
+   *     threshold. The first url in the list will be fetched. Recursive calls will always send in
+   *     all elements except the first one. Processing ends when there are no urls in the list.
+   * @param optionalPreviousRootNode The response from the previous findAddressCandidates call, if
+   *     any
+   * @return The CompletionStage for result of findAddressCandidates, if any. The `candidates` array
+   *     will be the merged results of multiple endpoints if more than one is called.
+   */
+  private CompletionStage<Optional<JsonNode>> processAddressSuggestionUrlsSequentially(
```

Comment:
suggestion (non-blocking): Add some unit tests for this method. It's pretty complex and I think it would help to have some tests to demonstrate how it works and guard against breaking changes. I could be convinced we don't need them though.

Reply:
I don't think this specifically needs tests. The `fetchAddressSuggestions` is calling this directly to kick off the recursion so those tests are already exercising it. Anything else would end up being duplicative. I'll see if I can expand on the existing tests though.